rust/otap-dataflow/benchmarks/benches/retry_processor/main.rs

```diff
@@ -0,0 +1,331 @@
+    message::{ControlMsg, Message, Sender},
+};
+use otap_df_otlp::grpc::OTLPData;
+use otap_df_otlp::proto::opentelemetry::collector::logs::v1::ExportLogsServiceRequest;
+use std::hint::black_box;
+use std::time::{Duration, Instant};
+use tokio::task::LocalSet;
+
+#[global_allocator]
+static GLOBAL: MiMalloc = MiMalloc;
+
+const LARGE_MSG_COUNT: usize = 1_000;
+const PENDING_MESSAGE_SCALES: &[usize] = &[100, 500, 1_000];
+const CLEANUP_SCALES: &[usize] = &[50, 100, 200];
+
+/// Simple mock effect handler that doesn't actually send messages
+struct MockEffectHandler<T> {
+    message_count: std::rc::Rc<std::cell::RefCell<usize>>,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> MockEffectHandler<T> {
+    fn new() -> Self {
+        Self {
+            message_count: std::rc::Rc::new(std::cell::RefCell::new(0)),
+            _phantom: std::marker::PhantomData,
+        }
+    }
+
+    #[allow(dead_code)]
+    fn get_count(&self) -> usize {
+        *self.message_count.borrow()
+    }
+
+    async fn send_message(&mut self, _data: T) -> Result<(), otap_df_engine::error::Error<T>> {
+        *self.message_count.borrow_mut() += 1;
+        Ok(())
+    }
+}
+
+/// Creates test OTLP logs data for benchmarking
+fn create_otlp_logs_data() -> OTLPData {
+    let logs_request = ExportLogsServiceRequest::default();
+    OTLPData::Logs(logs_request)
+}
+
+/// Benchmark 1: Message throughput comparison between success path and retry scenarios
+fn bench_message_throughput(c: &mut Criterion) {
+    let rt = tokio::runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .expect("failed to build Tokio runtime");
+
+    // Pin to a core for consistent measurements
+    let cores = core_affinity::get_core_ids().expect("couldn't get core IDs");
+    let core = cores.iter().last().expect("no cores found");
+    _ = core_affinity::set_for_current(*core);
+
+    let mut group = c.benchmark_group("retry_processor_throughput");
+
+    // Benchmark pure success path (0% retries)
+    let _ = group.throughput(Throughput::Elements(LARGE_MSG_COUNT as u64));
+    let _ = group.bench_function(BenchmarkId::new("success_path", LARGE_MSG_COUNT), |b| {
+        b.to_async(&rt).iter(|| async {
+            let local = LocalSet::new();
+
+            local
+                .run_until(async {
+                    let _processor = RetryProcessor::<OTLPData>::new();
+                    let mut effect_handler = MockEffectHandler::<OTLPData>::new();
+
+                    // Send OTLP logs messages (success path)
+                    for _i in 0..LARGE_MSG_COUNT {
+                        let otlp_data = create_otlp_logs_data();
+                        effect_handler
+                            .send_message(otlp_data)
+                            .await
+                            .expect("failed to send message in success path benchmark");
+
+                        // No need to simulate ACK/NACK for throughput test
```

Comment:
Since this benchmark group is trying to compare the 0% retry with 20% retry, shouldn't we still have the processor send process the request and also ACK the request in the benchmark?

If so, I wonder if we could simplify the benchmark by having the benchmark group for `with_20_percent_retries` parameterized by the percent of retries, and one of the benchmark cases could be 0%?

Reply:
Looks like this change was made. nice work!